indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/NumberEncoding.kt

```diff
@@ -241,222 +163,374 @@ fun Int.toUnsignedByteArray() = toLong().toUnsignedByteArray()
-    while (hasNext()) {
-        val curByte = next()
-        val current = BigInteger(curByte.toUByte().toInt())
-        accumulator += curByte
-        result = (current and mask) or (result shl 7)
-        if (current < 0x80.toUByte()) break
+private fun Sink.writeTwosComplement(number: Long, padded: Boolean = true): Int = when {
+    (number >= -0x80L && number <= 0x7FL) -> {
+        writeByte(number.toByte())
+        1
+    }
+
+    (number >= -0x8000L && number <= 0x7FFFL) -> {
+        val byteWritten = writeOrSkipPadding(number, 8, padded)
+        writeByte(number.toByte())
+        1 + byteWritten
+    }
+
+    (number >= -0x800000L && number <= 0x7FFFFFL) -> {
+        val byteWritten = writeOrSkipPadding(number, 16, padded)
+        writeByte((number ushr 8).toByte())
+        writeByte(number.toByte())
+        2 + byteWritten
+    }
+
+    (number >= -0x80000000L && number <= 0x7FFFFFFFL) -> {
+        val byteWritten = writeOrSkipPadding(number, 24, padded)
+        writeByte((number ushr 16).toByte())
+        writeByte((number ushr 8).toByte())
+        writeByte(number.toByte())
+        3 + byteWritten
+    }
+
+    (number >= -0x8000000000L && number <= 0x7FFFFFFFFFL) -> {
+        val byteWritten = writeOrSkipPadding(number, 32, padded)
+        writeByte((number ushr 24).toByte())
+        writeByte((number ushr 16).toByte())
+        writeByte((number ushr 8).toByte())
+        writeByte(number.toByte())
+        4 + byteWritten
+    }
+
+    (number >= -0x800000000000L && number <= 0x7FFFFFFFFFFFL) -> {
+        val byteWritten = writeOrSkipPadding(number, 40, padded)
+        writeByte((number ushr 32).toByte())
+        writeByte((number ushr 24).toByte())
+        writeByte((number ushr 16).toByte())
+        writeByte((number ushr 8).toByte())
+        writeByte(number.toByte())
+        5 + byteWritten
+    }
+
+    (number >= -0x80000000000000L && number <= 0x7FFFFFFFFFFFFFL) -> {
+        val byteWritten = writeOrSkipPadding(number, 48, padded)
+        writeByte((number ushr 40).toByte())
+        writeByte((number ushr 32).toByte())
+        writeByte((number ushr 24).toByte())
+        writeByte((number ushr 16).toByte())
+        writeByte((number ushr 8).toByte())
+        writeByte(number.toByte())
+        6 + byteWritten
+    }
+
+    else -> {
+        val byteWritten = writeOrSkipPadding(number, 56, padded)
+        writeByte((number ushr 48).toByte())
+        writeByte((number ushr 40).toByte())
+        writeByte((number ushr 32).toByte())
+        writeByte((number ushr 24).toByte())
+        writeByte((number ushr 16).toByte())
+        writeByte((number ushr 8).toByte())
+        writeByte(number.toByte())
+        7 + byteWritten
     }
 
-    return result to accumulator.toByteArray()
 }
 
+private inline fun Sink.writeOrSkipPadding(number: Long, shift: Int, padded: Boolean): Int {
+    val byte = (number ushr shift).toByte()
+    val writeFirstByte = padded || (byte != 0.toByte())
```

Comment:
why is this an `or`? should it not be `and`? as i understand the `padded` param to writevarint, it can be used to _suppress_ the leading zero (the leading zero should _only_ be written if it is both _needed_ and _padded = true_.)

Reply:
i also think that this is a hack, and would suggest either introducing a `writeUnsignedBytes` helper, or a lightweight `SkipLeadingZeroSinkWrapper` that does nothing except omit the first write if it is zero.

both approaches are far cleaner than this mess

Reply:
how would you craft such a wrapper, given we write each byte individually?

Reply:
not sure what the sink interface looks like offhand, but
```
class SinkWrapper(val wrapped: Sink) {
    var isFirst = True
    override fun writeByte(it: Byte) {
        val skip = (isFirst && (it == 0))
        isFirst = false;
        if(!skip) wrapped.writeByte(it)
    }
```

Reply:
reworked as per our offline discussion